### PR TITLE
Make BatchIterator killable

### PR DESCRIPTION
**Pull request by mfussenegger**
Initially we'd planned to create a "KillableRegistry" in order to avoid
having to implement kill on each BatchIterator or BatchConsumer
implementation. But currently adding a KillableRegistry would involve a
lot of work which is not practical as we're still in the migration from
RowReceiver from BatchIterator.

This changes BatchIterator to extend Killable as a programatic solution
which we may revisit later on.

Since BatchIterator is mostly a passiv component (except loadNextBatch),
kill won't actively do anything. Instead it causes further `moveNext()`
calls to fail.
This means for BatchConsumers it just looks like a regular failure -
from their perspective kill is no different to a regular error.

### CODE REVIEW

File: dex/src/main/java/io/crate/data/AsyncOperationBatchIterator.java [hunk 1]
```diff
@@ -67,6 +68,7 @@
     private boolean sourceExhausted = false;
     private boolean closed = false;
     private boolean loading = false;
+    private volatile Throwable killed;
```
Comment (seut) [inline dex/src/main/java/io/crate/data/AsyncOperationBatchIterator.java]: pls mark as nullable

File: dex/src/main/java/io/crate/data/CloseAssertingBatchIterator.java [hunk 2]
```diff
@@ -23,13 +23,16 @@
...
+import io.crate.exceptions.Exceptions;
 
+import javax.annotation.Nullable;
 import java.util.concurrent.CompletionStage;
 
 public class CloseAssertingBatchIterator implements BatchIterator {
 
     private final BatchIterator delegate;
     private boolean closed = false;
+    private volatile Throwable killed = null;
```
Comment (seut) [inline dex/src/main/java/io/crate/data/CloseAssertingBatchIterator.java]: maybe also mark as nullable

File: dex/src/main/java/io/crate/data/join/NestedLoopBatchIterator.java [hunk 3]
```diff
@@ -76,6 +78,7 @@ public static BatchIterator rightJoin(BatchIterator left,
     final CombinedColumn rowData;
     final BatchIterator left;
     final BatchIterator right;
+    private volatile Throwable killed;
```
Comment (seut) [inline dex/src/main/java/io/crate/data/join/NestedLoopBatchIterator.java]: unused?

File: sql/src/main/java/io/crate/operation/collect/BatchIteratorCollectorBridge.java [hunk 4]
```diff
@@ -36,61 +34,77 @@
...
+        private final Supplier<CompletableFuture<BatchIterator>> batchIteratorSupplier;
         private final BatchConsumer consumer;
-        private final Killable killable;
 
-        AsyncBatchItCollector(Supplier<CompletableFuture<BatchIterator>> batchIteratorFuture,
-                              BatchConsumer consumer,
-                              Killable killable) {
-            this.batchIteratorFuture = batchIteratorFuture;
+        private BatchIterator batchIterator;
+        private Throwable killed = null;
```
Comment (seut) [inline sql/src/main/java/io/crate/operation/collect/BatchIteratorCollectorBridge.java]: also here, maybe mark as nullable

File: sql/src/main/java/io/crate/operation/projectors/IndexWriterCountBatchIterator.java [hunk 5]
```diff
@@ -228,5 +227,11 @@ private void raiseIfLoading() {
             throw new IllegalStateException("Iterator is loading");
         }
     }
+
+    @Override
+    public void kill(@javax.annotation.Nullable Throwable throwable) {
```
Comment (seut) [inline sql/src/main/java/io/crate/operation/projectors/IndexWriterCountBatchIterator.java]: would prefer one @Nullable import per file

File: sql/src/main/java/io/crate/operation/collect/collectors/LuceneBatchIterator.java [hunk 6]
```diff
@@ -220,9 +222,17 @@ private void onDoc(int doc, LeafReader reader) throws IOException {
...
         if (closed) {
             throw new IllegalStateException("BatchIterator is closed");
         }
+        if (killed != null) {
+            Exceptions.rethrowUnchecked(killed);
+        }
+    }
+
+    @Override
+    public void kill(@Nullable Throwable throwable) {
```
Comment (andreidan) [inline sql/src/main/java/io/crate/operation/collect/collectors/LuceneBatchIterator.java]: Why is the throwable nullable for iterators? If there's no exceptional reason to kill an iterator close should be used right?

**Pull request merged**